src/compiler/checker.ts

```diff
@@ -3422,12 +3471,18 @@ module ts {
                         else {
                             type = getDeclaredTypeOfSymbol(symbol);
                             if (type.flags & (TypeFlags.Class | TypeFlags.Interface) && type.flags & TypeFlags.Reference) {
-                                let typeParameters = (<InterfaceType>type).typeParameters;
-                                if (node.typeArguments && node.typeArguments.length === typeParameters.length) {
-                                    type = createTypeReference(<GenericType>type, map(node.typeArguments, getTypeFromTypeNode));
+                                // In a type reference, the outer type parameters of the referenced class or interface are automatically
+                                // supplied as type arguments and the type reference only specifies arguments for the local type parameters

```

Comment:
Oh how interesting! I see how the cyclic instantiation happens.

Btw, I came up with another example of this, based on a combination of the example you just gave, and the one from #3237. It looks like this:
```ts
function foo<T>(x: T) {
    var z = foo(y);
    var y: {
        y1: T;
        y2: typeof z
    };
    return y;
}


function bar<T>(x: T) {
    var z = foo(y);
    var y: {
        y1: T;
        y2: typeof z;
    }
    return y;
}

var a = foo<string>("hello");
var b = bar<string>("hello");
a = b;
```
Seems to stack overflow for me. Does your change fix this?